File: Source/Client/Game/Objects/Item.vb
Imports System.IO
Imports Mirage.Sharp.Asfw
Imports Core

Module Item

#Region "Database"
    Friend Sub ClearItem(index As Integer)
        Type.Item(index) = Nothing
        For X = 0 To StatType.Count - 1
            ReDim Type.Item(index).Add_Stat(x)
        Next

        For X = 0 To StatType.Count - 1
            ReDim Type.Item(index).Stat_Req(x)
        Next

        Type.Item(index).Name = ""
        Type.Item(index).Description = ""
        GameState.Item_Loaded(index) = 0
    End Sub

    Sub ClearItems()
        Dim i As Integer

        ReDim Type.Item(MAX_ITEMS)

        For i = 1 To MAX_ITEMS
            ClearItem(i)
        Next

    End Sub
    
    Sub StreamItem(itemNum As Integer)
        If itemNum > 0 and Type.Item(itemNum).Name = "" Or GameState.Item_Loaded(itemNum) = 0 Then
            GameState.Item_Loaded(itemNum) = 1
            SendRequestItem(itemNum)
        End If
    End Sub

#End Region

#Region "Incoming Packets"

    Friend Sub Packet_UpdateItem(ByRef data() As Byte)
        Dim n As Integer, i As Integer
        Dim buffer As New ByteStream(data)

        n = buffer.ReadInt32

        ' Update the item
        Type.Item(n).AccessReq = buffer.ReadInt32()

        For i = 1 To StatType.Count - 1
            Type.Item(n).Add_Stat(i) = buffer.ReadInt32()
        Next

        Type.Item(n).Animation = buffer.ReadInt32()
        Type.Item(n).BindType = buffer.ReadInt32()
        Type.Item(n).JobReq = buffer.ReadInt32()
        Type.Item(n).Data1 = buffer.ReadInt32()
        Type.Item(n).Data2 = buffer.ReadInt32()
        Type.Item(n).Data3 = buffer.ReadInt32()
        Type.Item(n).TwoHanded = buffer.ReadInt32()
        Type.Item(n).LevelReq = buffer.ReadInt32()
        Type.Item(n).Mastery = buffer.ReadInt32()
        Type.Item(n).Name = buffer.ReadString()
        Type.Item(n).Paperdoll = buffer.ReadInt32()
        Type.Item(n).Icon = buffer.ReadInt32()
        Type.Item(n).Price = buffer.ReadInt32()
        Type.Item(n).Rarity = buffer.ReadInt32()
        Type.Item(n).Speed = buffer.ReadInt32()

        Type.Item(n).Stackable = buffer.ReadInt32()
        Type.Item(n).Description = buffer.ReadString()

        For i = 1 To StatType.Count - 1
            Type.Item(n).Stat_Req(i) = buffer.ReadInt32()
        Next

        Type.Item(n).Type = buffer.ReadInt32()
        Type.Item(n).SubType = buffer.ReadInt32

        Type.Item(n).KnockBack = buffer.ReadInt32()
        Type.Item(n).KnockBackTiles = buffer.ReadInt32()

        Type.Item(n).Projectile = buffer.ReadInt32()
        Type.Item(n).Ammo = buffer.ReadInt32()

        If n = GameState.descLastItem Then
            GameState.descLastType = 0
            GameState.descLastItem = 0
        End If

        buffer.Dispose()

    End Sub

#End Region

#Region "Outgoing Packets"

    Sub SendRequestItem(itemNum As Integer)
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CRequestItem)
        buffer.WriteInt32(itemNum)

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

#End Region

End Module

File: Source/Client/Game/Objects/Item.cs
using Core;
using Microsoft.VisualBasic.CompilerServices;
using Mirage.Sharp.Asfw;

namespace Client
{

    public class Item
    {

        #region Database
        public static void ClearItem(int index)
        {
            int statCount = System.Enum.GetValues(typeof(Stat)).Length;
                
            Core.Data.Item[index] = default;
            Array.Resize(ref Core.Data.Item[index].Add_Stat, statCount);
            Array.Resize(ref Core.Data.Item[index].Stat_Req, statCount);

            Core.Data.Item[index].Name = "";
            Core.Data.Item[index].Description = "";
            GameState.Item_Loaded[index] = 0;
        }

        public static void ClearItems()
        {
            int i;

            Core.Data.Item = new Core.Type.Item[Core.Constant.MAX_ITEMS];

            for (i = 0; i < Constant.MAX_ITEMS; i++)
                ClearItem(i);

        }

        public static void ClearChangedItem()
        {
            GameState.Item_Changed = new bool[Core.Constant.MAX_ITEMS];
        }

        public static void StreamItem(int itemNum)
        {
            if (itemNum >= 0 && string.IsNullOrEmpty(Core.Data.Item[itemNum].Name) && GameState.Item_Loaded[itemNum] == 0)
            {
                GameState.Item_Loaded[itemNum] = 1;
                SendRequestItem(itemNum);
            }
        }

        #endregion

        #region Incoming Packets

        public static void Packet_UpdateItem(ref byte[] data)
        {
            int n;
            int i;
            var buffer = new ByteStream(data);

            n = buffer.ReadInt32();

            // Update the item
            Core.Data.Item[n].AccessReq = buffer.ReadInt32();

            int statCount = System.Enum.GetValues(typeof(Stat)).Length;
            for (i = 0; i < statCount; i++)
                Core.Data.Item[n].Add_Stat[i] = (byte)buffer.ReadInt32();

            Core.Data.Item[n].Animation = buffer.ReadInt32();
            Core.Data.Item[n].BindType = (byte)buffer.ReadInt32();
            Core.Data.Item[n].JobReq = buffer.ReadInt32();
            Core.Data.Item[n].Data1 = buffer.ReadInt32();
            Core.Data.Item[n].Data2 = buffer.ReadInt32();
            Core.Data.Item[n].Data3 = buffer.ReadInt32();
            Core.Data.Item[n].LevelReq = buffer.ReadInt32();
            Core.Data.Item[n].Mastery = (byte)buffer.ReadInt32();
            Core.Data.Item[n].Name = buffer.ReadString();
            Core.Data.Item[n].Paperdoll = buffer.ReadInt32();
            Core.Data.Item[n].Icon = buffer.ReadInt32();
            Core.Data.Item[n].Price = buffer.ReadInt32();
            Core.Data.Item[n].Rarity = (byte)buffer.ReadInt32();
            Core.Data.Item[n].Speed = buffer.ReadInt32();

            Core.Data.Item[n].Stackable = (byte)buffer.ReadInt32();
            Core.Data.Item[n].Description = buffer.ReadString();

            for (i = 0; i < statCount; i++)
                Core.Data.Item[n].Stat_Req[i] = (byte)buffer.ReadInt32();

            Core.Data.Item[n].Type = (byte)buffer.ReadInt32();
            Core.Data.Item[n].SubType = (byte)buffer.ReadInt32();

            Core.Data.Item[n].KnockBack = (byte)buffer.ReadInt32();
            Core.Data.Item[n].KnockBackTiles = (byte)buffer.ReadInt32();

            Core.Data.Item[n].Projectile = buffer.ReadInt32();
            Core.Data.Item[n].Ammo = buffer.ReadInt32();

            if (n == GameState.descLastItem)
            {
                GameState.descLastType = 0;
                GameState.descLastItem = 0L;
            }

            buffer.Dispose();

        }

        #endregion

        #region Outgoing Packets

        public static void SendRequestItem(int itemNum)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CRequestItem);
            buffer.WriteInt32(itemNum);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        #endregion

    }
}